krm-functions/nad-fn/mutator/mutator.go

```diff
@@ -0,0 +1,179 @@
+	clusterctxtlibv1alpha1 "github.com/nephio-project/nephio/krm-functions/lib/clustercontext/v1alpha1"
+	interfacelibv1alpha1 "github.com/nephio-project/nephio/krm-functions/lib/interface/v1alpha1"
+	//ipallocv1v1alpha1 "github.com/nephio-project/nephio/krm-functions/lib/ipallocation/v1alpha1"
+	condkptsdk "github.com/nephio-project/nephio/krm-functions/lib/condkptsdk"
+	ipalloclibv1alpha1 "github.com/nephio-project/nephio/krm-functions/lib/ipallocation/v1alpha1"
+	nadlibv1 "github.com/nephio-project/nephio/krm-functions/lib/nad/v1"
+	ipamv1alpha1 "github.com/nokia/k8s-ipam/apis/alloc/ipam/v1alpha1"
+	vlanv1alpha1 "github.com/nokia/k8s-ipam/apis/alloc/vlan/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type mutatorCtx struct {
+	fnCondSdk       condkptsdk.KptCondSDK
+	masterInterface string
+	cniType         string
+	siteCode        string
+}
+
+func Run(rl *fn.ResourceList) (bool, error) {
+	m := mutatorCtx{}
+	var err error
+	m.fnCondSdk, err = condkptsdk.New(
+		rl,
+		&condkptsdk.Config{
+			For: corev1.ObjectReference{
+				APIVersion: nadv1.SchemeGroupVersion.Identifier(),
+				Kind:       reflect.TypeOf(nadv1.NetworkAttachmentDefinition{}).Name(),
+			},
+			Watch: map[corev1.ObjectReference]condkptsdk.WatchCallbackFn{
+				{
+					APIVersion: infrav1alpha1.GroupVersion.Identifier(),
+					Kind:       reflect.TypeOf(infrav1alpha1.ClusterContext{}).Name(),
+				}: m.ClusterContextCallbackFn,
+				{
+					APIVersion: ipamv1alpha1.GroupVersion.Identifier(),
+					Kind:       ipamv1alpha1.IPAllocationKind,
+				}: nil,
+				{
+					APIVersion: vlanv1alpha1.GroupVersion.Identifier(),
+					Kind:       vlanv1alpha1.VLANAllocationKind,
+				}: nil,
+				{
+					APIVersion: nephioreqv1alpha1.GroupVersion.Identifier(),
+					Kind:       nephioreqv1alpha1.InterfaceKind,
+				}: nil,
+			},
+			PopulateOwnResourcesFn: nil,
+			GenerateResourceFn:     m.generateResourceFn,
+		},
+	)
+	if err != nil {
+		rl.Results = append(rl.Results, fn.ErrorResult(err))
+	}
+	return m.fnCondSdk.Run()
+}
+
+func (r *mutatorCtx) ClusterContextCallbackFn(o *fn.KubeObject) error {
+	clusterContext := clusterctxtlibv1alpha1.NewMutator(o.String())
+	cluster, err := clusterContext.UnMarshal()
+	if err != nil {
+		return err
+	}
+	if cluster.Spec.CNIConfig.MasterInterface == "" {
+		return fmt.Errorf("MasterInterface on ClusterContext cannot be empty")
+	} else {
+		r.masterInterface = cluster.Spec.CNIConfig.MasterInterface
+	}
+	if cluster.Spec.CNIConfig.CNIType == "" {
+		return fmt.Errorf("CNIType on ClusterContext cannot be empty")
+	} else {
+		r.cniType = cluster.Spec.CNIConfig.CNIType
+	}
+	if cluster.Spec.SiteCode == nil {
+		return fmt.Errorf("SiteCode on ClusterContext cannot be empty")
+	} else {
+		r.siteCode = *cluster.Spec.SiteCode
+	}
+	return nil
+}
+
+func (r *mutatorCtx) generateResourceFn(forObj *fn.KubeObject, objs fn.KubeObjects) (*fn.KubeObject, error) {
+	if len(objs) == 0 || len(objs) < 2 {
```

Comment:
This is a partial check and doesn't serve much usefulness I feel. Firstly because, we don't know what object is missing. NAD needs a pair of IPAlloc and VLANAlloc. I would recommend check if all the named resource have that pair present using `map[string][]any` or similar. Because this check would fail if there are two IPAlloc and not VLANAlloc at all present.